perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_institutionsCountries.java

```diff
@@ -107,44 +108,69 @@ public String modifyValue(PerunSession session, DnsMapCtx ctx, UserExtSource ues
-		List<String> messages = super.resolveVirtualAttributeValueChange(sess, message);
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl sess, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> messages = super.resolveVirtualAttributeValueChange(sess, message);
 		// react to auditlog message about changing dnsStateMapping
-		String key = null;
-		Matcher setMatcher = AUDIT_MESSAGE_SET_DNS_DOMAIN_COUNTRY.matcher(message);
-		if(setMatcher.matches()) {
-			key = setMatcher.group(3);
-		} else {
-			Matcher delMatcher = AUDIT_MESSAGE_REMOVED_DNS_DOMAIN_COUNTRY.matcher(message);
-			if(delMatcher.matches()) {
-				key = delMatcher.group(3);
-			}
+
+		if (message instanceof AttributeSetForKey
+			&& ((AttributeSetForKey) message).getAttribute().getFriendlyName().equals(DNS_STATE_MAPPING_ATTR.getFriendlyName())) {
+
+			messages.addAll(resolveAttributeSetForKey(sess, (AttributeSetForKey) message));
+		} else if (message instanceof AttributeRemovedForKey
+			&& ((AttributeRemovedForKey) message).getAttribute().getFriendlyName().equals(DNS_STATE_MAPPING_ATTR.getFriendlyName())) {
+
+			messages.addAll(resolveAttributeRemovedForKey(sess, (AttributeRemovedForKey) message));
 		}
-		if (key!=null) {
-			//find longer matching domains, e.g. ics.muni.cz and muni.cz for .cz
-			List<String> longerDomains =  new ArrayList<>();
-			AttributesManagerBl am = sess.getPerunBl().getAttributesManagerBl();
-			for(String dnsDomain : am.getEntitylessStringAttributeMapping(sess, DNS_STATE_MAPPING_ATTR.getName()).keySet()) {
-				if(dnsDomain.endsWith(key) && dnsDomain.length()>key.length()) {
-					longerDomains.add(dnsDomain);
-				}
-			}
-			log.debug("DNS domain '{}' changed, found longer domains: {}",key, longerDomains);
-			//find users that are affected by the change - have schacHomeOrganization value ending in key, but not ending with longerDomains
-			List<User> affectedUsers = sess.getPerunBl().getUsersManagerBl().findUsersWithExtSourceAttributeValueEnding(sess,getSourceAttributeName(),key,longerDomains);
-			for (User user : affectedUsers) {
-				Attribute thisAttribute = am.getAttribute(sess, user, getDestinationAttributeName());
-				messages.add(thisAttribute.serializeToString() + " set for " + user.serializeToString() + ".");
+		return messages;
+	}
+
+	private List<AuditEvent> resolveAttributeSetForKey(PerunSessionImpl sess, AttributeSetForKey attributeSetForKey) throws WrongAttributeAssignmentException, InternalErrorException, AttributeNotExistsException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+
+		String key = attributeSetForKey.getKey();
+		List<String> longerDomains =  new ArrayList<>();
```

Comment:
You can extract the code from this method into another private method since there is exactly the same code in the method resolveAttributeRemovedForKey.